app/src/main/scala/org/alephium/explorer/service/TokenSupplyService.scala

```diff
@@ -106,30 +143,35 @@ object TokenSupplyService {
       })
     }
 
-    private val mainInputs  = InputSchema.table.filter(_.mainChain)
-    private val mainOutputs = OutputSchema.table.filter(_.mainChain)
-
-    private def genesisLockedToken(lockTime: TimeStamp) = {
-      mainOutputs
-        .join(
-          mainTransactions
-            .join(BlockHeaderSchema.table.filter(_.height === Height.unsafe(0)))
-            .on(_.blockHash === _.hash)
-        )
-        .on(_.txHash === _._1.hash)
-        .filter { case (output, _) => output.lockTime.map(_ > lockTime).getOrElse(false) }
-        .map { case (output, _) => output.amount }
-        .sum
+    private def circulatingTokensQuery(at: TimeStamp) = {
+      sql"""
+      SELECT sum(outputs.amount)
+      FROM outputs
+      LEFT JOIN inputs
+        ON outputs.key = inputs.output_ref_key
+        AND inputs.main_chain = true
+        AND inputs.block_timestamp <= $at
+      WHERE outputs.block_timestamp >= ${ALPH.LaunchTimestamp} /* We don't count genesis address */
+      AND outputs.block_timestamp <= $at
+      AND (outputs.lock_time is NULL OR outputs.lock_time <= $at) /* Only count unlock tokens */
+      AND outputs.main_chain = true
+      AND outputs.address NOT IN #$excludedAddresses /* We exclude the reserved wallets */
+      AND inputs.block_hash IS NULL;
+        """.as[Option[U256]].exactlyOne
     }
 
-    private def unspentTokens(at: TimeStamp) = {
-      mainOutputs
-        .filter(_.timestamp <= at)
-        .joinLeft(mainInputs.filter(_.timestamp <= at))
-        .on(_.key === _.outputRefKey)
-        .filter { case (_, inputOpt) => inputOpt.isEmpty }
-        .map { case (output, _) => output.amount }
-        .sum
+    private def allUnspentTokens(at: TimeStamp) = {
+      sql"""
+        SELECT sum(outputs.amount)
+        FROM outputs
+        LEFT JOIN inputs
+          ON outputs.key = inputs.output_ref_key
+          AND inputs.main_chain = true
+          AND inputs.block_timestamp <= $at
+        WHERE outputs.main_chain = true
+        AND outputs.block_timestamp <= $at
+        AND inputs.block_hash IS NULL;
+      """.as[U256].exactlyOne
```

Comment:
Hey Thomas, I got `NullPointerException` on empty table for running this query. I'm sure being `private` it is safe to call during runtime.

> java.lang.RuntimeException, with message: java.lang.NullPointerException: Cannot invoke "scala.math.BigDecimal.toBigInt()" because the return value of "slick.jdbc.PositionedResult.nextBigDecimal()" is null.
...

Test case
```scala
it should "empty" in {
  val tokenSupplyService =
    new TokenSupplyService.Impl(Duration.unsafe(1), databaseConfig, 1)

  run(tokenSupplyService.allUnspentTokens(TimeStamp.now())).futureValue
}
```
Should we still return `Option[U256]` here anyway so that this function is total and we don't get the `NullPointerException` (the same we saw last week)?

```scala
private def allUnspentTokensOption(at: TimeStamp): DBActionR[Option[U256]] = ...

private def allUnspentTokens(at: TimeStamp): DBActionR[U256] =
  allUnspentTokensOption(at).map(_.getOrElse(U256.Zero))
```

Reply:
ooooh of course :sweat_smile:  nice catch

Reply:
thx :+1: I fixed in a new commit.